controllers/topology/conditions.go

```diff
@@ -0,0 +1,116 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topology
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/scope"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+func (r *ClusterReconciler) reconcileConditions(s *scope.Scope, cluster *clusterv1.Cluster, reconcileErr error) error {
+	return r.reconcileTopologyReconciledCondition(s, cluster, reconcileErr)
+}
+
+// reconcileTopologyReconciledCondition sets the TopologyReconciled condition on the cluster.
+// The TopologyReconciled condition is considered true is spec of all the objects associated with the
+// cluster are in sync with the topology defined in the cluster.
+// The condition is false under the following conditions:
+// - An error occurred during the reonciled process of the cluster topology.
+// - The cluster upgrade has not yet propogated to all the components of the cluster.
+//   - For a managed topology cluster the version upgrade is propagated one component at a time.
+//     In such a case, since some of the component's spec would be adrift from the topology the
+//     topology cannot be considered fully reconciled.
+func (r *ClusterReconciler) reconcileTopologyReconciledCondition(s *scope.Scope, cluster *clusterv1.Cluster, reconcileErr error) error {
+	// If an erorr occurred during reconciliation set the TopologyReconciled condition to false.
+	// Add the error message from the reoncile function to the message of the condition.
+	if reconcileErr != nil {
+		conditions.Set(
+			cluster,
+			conditions.FalseCondition(
+				clusterv1.TopologyReconciledCondition,
+				clusterv1.TopologyReconciledErroredReason,
+				clusterv1.ConditionSeverityError,
+				reconcileErr.Error(),
+			),
+		)
+		return nil
+	}
+
+	// If either the Control Plane or any of the MachineDeployments are still pending to pick up the new version (generally
+	// happens when upgrading the cluster) then the topology is not considered as fully reconciled.
+	if s.UpgradeTracker.ControlPlane.PendingUpgrade || s.UpgradeTracker.MachineDeployments.PendingUpgrade() {
+		var msg string
+		var reason string
+		if s.UpgradeTracker.ControlPlane.PendingUpgrade {
+			msg = fmt.Sprintf("Control plane upgrade to %s on hold. ", s.Blueprint.Topology.Version)
+			reason = clusterv1.TopologyReconciledControlPlaneUpgradePendingReason
+		} else {
+			msg = fmt.Sprintf("MachineDeployment(s) %s upgrade to version %s on hold. ",
+				strings.Join(s.UpgradeTracker.MachineDeployments.DeferredUpgradeNames(), ", "),
+				s.Blueprint.Topology.Version,
+			)
+			reason = clusterv1.TopologyReconciledMachineDeploymentsUpgradePendingReason
+		}
+
+		switch {
+		case s.UpgradeTracker.ControlPlane.IsProvisioning:
+			msg += "Control Plane is in the process of completing initial provisioning"
```

Comment:
```suggestion
			msg += "Control plane is in the process of completing initial provisioning"
```
nit, to keep the spelling the same with the other occurrences